File: python/890 Find and Replace Pattern.py
class Solution:
    def findAndReplacePattern(self, words, pattern):
        dic = {}
        tmp = []
        count = 1
        res = []
        for i in pattern:
            if i in dic:
                tmp.append(dic[i])
            else:
                tmp.append(count)
                dic[i] = count
                count += 1
        
        dic2 = {}
        tmp2 = []
        count = 1
        for i in words:
            for j in i:
                if j in dic2:
                    tmp2.append(dic2[j])
                else:
                    tmp2.append(count)
                    dic2[j] = count
                    count += 1
                          
            if tmp == tmp2:
                res.append(i)
            count = 1
            dic2 = {}
            tmp2 = []
            
        return res

File: java/890 Find and Replace Pattern.java
class Solution {
    public List<String> findAndReplacePattern(String[] words, String pattern) {
        int n = pattern.length();
        List<String> res = new ArrayList<String>();

        HashMap<Character, Integer> dic = new HashMap<Character, Integer>();
        int count = 1;
        String tmp="";
        for (int i = 0; i < n; i++) {
            if (dic.get(pattern.charAt(i)) != null) {
                tmp += dic.get(pattern.charAt(i));
            }else {
                tmp += Integer.toString(count);
                dic.put(pattern.charAt(i), count);
                count += 1;
            }
        }

        HashMap<Character, Integer> dic2 = new HashMap<Character, Integer>();
        int count2 = 1;
        String tmp2= "";
        for (String i : words) {
            for (int j = 0; j < i.length(); j++) {
                if (dic2.get(i.charAt(j)) != null ) {
                    tmp2 += dic2.get(i.charAt(j));
                }else {
                    tmp2 += Integer.toString(count2);
                    dic2.put(i.charAt(j), count2);
                    count2 += 1;
                }
            }
            if (tmp.equalsIgnoreCase(tmp2)) {
                res.add(i);
            }
            count2 = 1;
            dic2.clear();
            tmp2 = "";
        }
        return res;
    }
}
